### DISCUSSION TITLE
Fully typed field presence check

### DISCUSSION THREAD

**Discussion by francipvb** [pydantic/pydantic#5579] [Question] [answered]
Hello,

I have a problem where a pydantic model is used to update an entity. However, I want to check if a field has been provided in a type-checked way, I mean, with future refactoring and errors reporting support.

I thought about something like this:

```python
from pydantic import BaseModel, is_present

class Model(BaseModel):
    a: int|None=None
    b: float
    c: int|None

class B(BaseModel):
    x: int=5

a=A(b=45.5, c=None)
is_present(A.a, a) # -> False
is_present(A.c, a) # -> True
b=B(x=5)
c=B()
is_present(B.x, c) # -> False
is_present(A.a, c) # -> Error!

```

Is there a solution for this, I mean a type-safe way?

**Comment by dmontagu** [comment]
We have the `__fields_set__` attribute which shows which fields were specified during validation. Does that help?

**Reply by francipvb** [reply]
This helps, but it is not type-safe. I mean, if I rename a field I am checking, the presence through the `__fields_set__` attribute will always fail.

Note that tools like pylance refactoring actions can rename attribute references, but the check you propose cann't be changed by this way.

Why the fields are not also descriptors? This would simplify this.

**Reply by dmontagu** [reply, marked as answer]
I think the best way to do this without a lot of involved changes would be to add a type called `Undefined` or similar, and set a value of it as the default. That way, you'll know it was set to that if and only if the default value was used. This has some rough edges but I think  should be possible to use in a type-safe way.

If you have issues that that doesn't address or otherwise want changes to the behavior (potentially including by using the fields as descriptors), I would suggest creating a new issue with a feature request, demonstrating the APIs you would want supported.

Given this hasn't been a common request over the years, I think it's more likely to get attention if you can provide a suggested implementation (which won't have significant negative performance implications for typical user code).

**Reply by francipvb** [reply]
Hello,

Ok, I'll do this.

Your proposal is good in some cases, but if you apply it to FastAPI, for example, you'll get errors in the OpenAPI generator.

Thank you, I will propose an implementation.
